tests/test_connect.py

```diff
@@ -48,53 +48,53 @@ def test_connect(self):
             self.fail(f'invalid value for {TEST_SERVER_ENV_VAR}')
 
         remove_existing_config()
-        app = create_test_app()
-        self.assertReachesInterfaceState(app, interface_state.ConfigurePredefinedServer)
-        self.assertIsNone(app.interface_state.results)
-        self.assertReachesNetworkState(app, network_state.UnconnectedState)
-
-        # enter the server address
-        app.interface_transition('enter_custom_address', test_server['address'])
-        self.assertReachesInterfaceState(app, interface_state.ConfigureCustomServer)
-        self.assertEqual(app.interface_state.address, test_server['address'])
-
-        # perform the oauth login
-        server = CustomServer(test_server['address'])
-        with patch('webbrowser.open') as webbrowser_open:
-            app.interface_transition('connect_to_server', server)
-            self.assertReachesInterfaceState(app, interface_state.OAuthSetup)
-            webbrowser_open.assert_called_once()
-            url = webbrowser_open.call_args[0][0]
-
-        # get the login form so we know where to post the data
-        session = requests.Session()
-        response = session.get(url)
-        self.assertIn('<h1>Sign In</h1>', response.text)
-        self.assertIn('name="userName"', response.text)
-        self.assertIn('name="userPass"', response.text)
-        self.assertIn('name="_form_auth_redirect_to"', response.text)
-        action = get_form_action(response.text)
-        self.assertIsNotNone(action)
-        self.assertTrue(action.startswith('/'))
-        auth_url = replace_url_path(response.url, action)
-
-        # post the login credentials
-        response = session.post(auth_url, data=dict(
-            userName=test_server['username'],
-            userPass=test_server['password'],
-            _form_auth_redirect_to=url,
-        ))
-        if 'The credentials you provided were not correct.' in response.text:
-            self.fail("invalid credentials for test server")
-        self.assertIn('<h1>Approve Application</h1>', response.text)
-
-        # post the approval for the eduvpn app
-        response = session.post(response.url, data=dict(approve='yes'))
-        self.assertIn('You can now close this window', response.text)
-
-        # now wait for the connection to be established
-        self.assertReachesInterfaceState(app, interface_state.ConnectionStatus)
-        self.assertReachesNetworkState(app, network_state.ConnectedState)
-
-        app.network_transition('disconnect')
-        self.assertReachesNetworkState(app, network_state.DisconnectedState)
+        with create_test_app() as app:
+            self.assertReachesInterfaceState(app, interface_state.ConfigurePredefinedServer)
```

Comment:
just a non-blocking style thing, I myself prefer minimal indentation, so if you have context managers with large amounts of code in it, I prefer to create a helper function. This improves readability and prevents mistakes.

Reply:
I agree on the indentation, but a function would also require try-finally and makes it too easy to forget to properly cleanup the thread that it runs. This is precisely what caused the test to hang. I could use unittest.setUp/tearDown or use a decorator but I find this more idiomatic and clean.